src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java

```diff
@@ -730,23 +1024,40 @@ static List<CompactionAggregate> getSelection(List<CompactionAggregate.UnifiedAg
                 continue; // compaction is too large for current cycle
 
             int currentLevel = levelOf(pick);
-            assert currentLevel >= 0 : "Invalid level in " + pick + ": level -1 is only allowed for expired-only compactions";
-
             boolean isAdaptive = controller.isRecentAdaptive(pick);
+            // avoid computing sharding stats if are not going to schedule the compaction at all
+            if (!reservations.hasRoom(currentLevel))
+                continue;
             if (isAdaptive && remainingAdaptiveCompactions <= 0)
-                continue; // do not allow more than remainingAdaptiveCompactions to limit latency spikes upon changing W
-
+                continue;
             if (shouldCheckSSTableSelected && !Collections.disjoint(selectedSSTables, pick.sstables()))
                 continue; // do not allow multiple selections of the same sstable
 
-            if (!reservations.accept(currentLevel))
+            int parallelism = controller.parallelizeOutputShards() ? getShardingStats(aggregate).coveredShardCount : 1;
+            if (parallelism > remaining)
+                parallelism = remaining;
+            assert currentLevel >= 0 : "Invalid level in " + pick;
+
+            if (isAdaptive)
+            {
+                if (parallelism > remainingAdaptiveCompactions)
+                {
+                    parallelism = remainingAdaptiveCompactions;
+                    if (parallelism <= 0)
+                        continue; // do not allow more than remainingAdaptiveCompactions to limit latency spikes upon changing W
+                }
+            }
+
+            parallelism = reservations.accept(currentLevel, parallelism);
+            if (parallelism <= 0)
```

Comment:
Nit: I don't think this can happen anymore, since we don't get here if `!hasRoom(currentLevel)` (at least with my understanding of `hasRoom` as meaning "will `accept` with a parameter > 0 give a result that is > 0?"). Not that the condition hurts and I don't mind keeping it, but maybe just worth adapting the comment and/or transforming into an assertion (if my reading here is correct obviously).

Reply:
Changed (also for the adaptive case above).